### DISCUSSION TITLE
make ".html" extension customizable

### DISCUSSION THREAD

**Discussion by Jc-L** [mkdocs/mkdocs#2967] [Ideas]
I am deploying on a platform that require a specific extension to the files to be served correctly, and that extension is not `.html`.
Would an option to configure the extension be merged if I provide the code or not (in which case I'll do a dirty parsing and forget about it) ?

Changes are mostly located in `structure/files.py`.

**Comment by jamesprnich** [comment]
I've just been looking for ways to allow me to do this. Would prefer to have it built into mkdocs than have to modify post build. 

Having this feature added and merged in would be awesome :)

**Comment by jamesprnich** [comment]
This might save others a little time....

Rename all *.html to *.aspx
`find . -name "*.html" -exec bash -c 'mv "$1" "${1%.html}".aspx' - '{}' \;`

Update all links to use the files with *.aspx extension
`find . -name "*.aspx" -exec sed -i 's/\.html/.aspx/g' {} \;`

Update the search json to use the *.aspx files
`find . -name "*.json" -exec sed -i 's/\.html/.aspx/g' {} \;`

**Reply by Jc-L** [reply]
Be aware that `find . -name "*.aspx" -exec sed -i 's/\.html/.aspx/g' {} \;` will also tranform any matching text outside link, as well as links towards any external website.
 
You can use something like
```
sed 's/\.html\(["\x27]\)/.aspx\1/g'
```
to  add a match on single or double quotes. Use a leading pattern like `\(href=["\x27]\/\/http:\/\/www\.mysite.*\)\.html\(["\x27]\)/\1.aspx\2` etc. to match more precisely your site, and replace the `.*` with the characters used in your urls. On the overall, this is prone to errors, hence the proposal to have something built-in.

**Comment by oprypin** [comment]
The proper current solution to replace the file extensions is the following:

```python
import re

def replace_extension(filename):
    return re.sub(r'\.html$', '.aspx', filename)

def on_files(files, **kwargs):
    for f in files:
        if f.is_documentation_page():
            f.dest_path = replace_extension(f.dest_path)
            f.abs_dest_path = replace_extension(f.abs_dest_path)
            f.url = replace_extension(f.url)
```

You can save this file as `hooks.py`
and add to the config:

```yaml
plugins:
  - mkdocs-simple-hooks:
      hooks:
        on_files: 'hooks:on_files'
```

This will make the build self-contained, to not need post-processing.
*Unfortunately*, it doesn't play well with `mkdocs serve`.

**Reply by sanzoghenzo** [reply]
Thanks, I was just trying to do exactly this!

as per `mkdocs-simple-hooks` documentation, one can use an environment variable to enable/disable the hooks:

```yaml
plugins:
  #...
  - mkdocs-simple-hooks:
      enabled: !ENV [ENABLE_MKDOCS_ASPX, False]
      hooks:
        on_files: "docs.hooks:on_files"
```

This way I can `mkdoc serve` html pages and build aspx files with

```shell
ENABLE_MKDOCS_ASPX=True mkdocs build
```

@Jc-L , are you by any chance trying to deploy to SharePoint site pages? if so, how did you manage this?
I would like to do that, but I didn't find a good solution to this (the only guide I found involves changing administrative settings, and I don't have those privileges).

I found the doctor (nodejs) project that creates a SharePoint page for each file and adds the markdown as a web part, but I would like to leverage the MkDocs ecosystem (automatic API reference creation for starters)... 
Looking at the documentation, it seems unlikely that I can create a plugin/hook to do what I want!

**Reply by Jc-L** [reply]
@sanzoghenzo, I was indeed trying to deploy on Sharepoint, but could not fully automate the deployment process. Currently I generate the mkdocs pages locally on a Windows laptop (with the `.aspx` extensions), and copy the resulting files to a Sharepoint location mounted through OneDrive. So still manual but working in my specific situation.

**Comment by Jc-L** [comment]
I found back a similar hook on `pre_page`. Not sure how it compares but it may help :

in the config file:
```
plugins:
  - mkdocs-simple-hooks:
      hooks:
        on_files: 'hooks:on_pre_page'
```

to add in a `hooks.py` file:
```python
import os.path
import posixpath

def on_pre_page(page, config, files):
    new_ext = '.aspx'
    file = page.file

    file.dest_path = os.path.splitext(file.dest_path)[0] + new_ext
    file.abs_dest_path = os.path.splitext(file.abs_dest_path)[0] + new_ext

    if file.url.endswith('/'):
        file.url += 'index' + new_ext
    else:
        file.url = posixpath.splitext(file.url)[0] + new_ext
```
